Revamped RSS instructions for content collections (#2401)
withastro/docs: src/pages/en/guides/rss.mdx

```

### Using glob imports

To create an RSS feed from documents in `src/pages/`, use the `pagesGlobToRssItems()` helper. This accepts an [`import.meta.glob`](https://vitejs.dev/guide/features.html#glob-import) result and outputs an array of valid RSS feed items (see [more about writing glob patterns](/en/guides/imports/#glob-patterns) for specifying which pages to include).

This function assumes, but does not verify, that all necessary feed properties are present in each document's frontmatter. If you encounter errors, verify each page frontmatter manually.

```js title="src/pages/rss.xml.js" "pagesGlobToRssItems" "await pagesGlobToRssItems("
import rss, { pagesGlobToRssItems } from '@astrojs/rss';

export async function get(context) {
  return rss({
    title: 'Buzz’s Blog',
    description: 'A humble Astronaut’s guide to the stars',
    site: context.site,
    items: await pagesGlobToRssItems(
      import.meta.glob('./blog/*.{md,mdx}'),
    ),
  });
}
```

### Including full post content

<Since v="1.6.14" />

The `content` key contains the full content of the post as HTML. This allows you to make your entire post content available to RSS feed readers.

:::tip

> Whenever you're using HTML content in XML, suggest using a package like [`sanitize-html`](https://www.npmjs.com/package/sanitize-html) in order to make sure that your content is properly sanitized, escaped, and encoded.

Reviewer: ```suggestion
Whenever you're using HTML content in XML, we suggest using a package like [`sanitize-html`](https://www.npmjs.com/package/sanitize-html) in order to make sure that your content is properly sanitized, escaped, and encoded.
```

Orr, if you were intentionally trying to "un-we" the voice of the tip, you could go for :

> Whenever you're using HTML content in XML, it's good practice to use... 

> Whenever you're using HTML content in XML, use

:::

When using content collections, render the post `body` using a standard Markdown parser like [`markdown-it`](https://github.com/markdown-it/markdown-it) and sanitize the result:

```js title="src/pages/rss.xml.js" ins={2, 3, 4, 14}
import rss from '@astrojs/rss';
import sanitizeHtml from 'sanitize-html';
import MarkdownIt from 'markdown-it';
const parser = new MarkdownIt();

export async function get(context) {
  const blog = await getCollection('blog');
  return rss({
    title: 'Buzz’s Blog',
    description: 'A humble Astronaut’s guide to the stars',
    site: context.site,
    items: blog.map((post) => ({
      link: `/blog/${post.slug}/`,
      // Note: this will not process components or JSX expressions in MDX files.
      content: sanitizeHtml(parser.render(post.body)),
      ...post.data,
    })),
  });
}
```

> When using glob imports with Markdown, we suggest using the `compiledContent()` helper to retrieve the rendered HTML for sanitization. Note this feature is **not** supported for MDX files.

Reviewer: ```suggestion
When using glob imports with Markdown, we suggest using the `compiledContent()` helper to retrieve the rendered HTML for sanitization. Note: this feature is **not** supported for MDX files.
```

### Using content collections

You may use [content collections](/en/guides/content-collections/) to manage your Markdown or MDX content. This exposes a `getCollection()` function to retrieve a list of items that you can map to an RSS feed.

For example, say you have a `blog` collection under `src/content/blog/`. You can retrieve this collection and generate an RSS feed like so:

```js title="src/pages/rss.xml.js" "items:" "const blog = await getCollection('blog');"
import rss from '@astrojs/rss';
import { getCollection } from 'astro:content';

export async function get(context) {
  const blog = await getCollection('blog');
  return rss({
    title: 'Buzz’s Blog',
    description: 'A humble Astronaut’s guide to the stars',
    site: context.site,
    items: blog.map((post) => ({
      title: post.data.title,
      pubDate: post.data.pubDate,
      description: post.data.description,
      customData: post.data.customData,
      // Compute RSS link from post `slug`
      // This example assumes all posts are rendered as `/blog/[slug]` routes
      link: `/blog/${post.slug}/`,
    })),
  });
}
```

:::tip

> The `link` property of each feed item can be computed from the content entry `slug`. See [our guide on generating pages from content collections](https://docs.astro.build/en/guides/content-collections/#generating-pages-from-content-collections) to see how slugs correspond to routes.

Reviewer: ```suggestion
The `link` property of each feed item can be computed from the content entry `slug`. See [our guide on generating pages from content collections](/en/guides/content-collections/#generating-pages-from-content-collections) to see how slugs correspond to routes.
```

:::

> `@astrojs/rss` also exposes an `rssSchema` for use with collection schemas. This ensures your frontmatter contains all properties expected by an RSS feed item. You can apply this schema to your content config like so:
> 
> ```js title="src/content/config.ts" "rssSchema"
> import { rssSchema } from '@astrojs/rss';
> 
> const blog = defineCollection({
>   schema: rssSchema,
> });
> 
> export const collections = { blog };

Reviewer: ```suggestion
You can configure your collection schema to enforce these expected RSS properties. Import and apply `rssSchema` to ensure that each collection entry produces a valid RSS feed item.

```js title="src/content/config.ts" "rssSchema"
import { rssSchema } from '@astrojs/rss';

const blog = defineCollection({
  schema: rssSchema,
});

export const collections = { blog };
```
Is it fair to say that this is the main point? I think something like this is more user-focused: Want this? Do this!

```

### Using glob imports

> You may generate an RSS feed from documents in `src/pages/`. Since `Astro.glob` is not available from endpoint routes, we've included a `pagesGlobToRssItems()` helper. This accepts an `import.meta.glob` result ([see Vite documentation](https://vitejs.dev/guide/features.html#glob-import)) and outputs an array of valid RSS feed items.

Reviewer: ```suggestion
To create an RSS feed from documents in `src/pages/`, use the `pagesGlobToRssItems()` helper. This accepts an [`import.meta.glob`](https://vitejs.dev/guide/features.html#glob-import) result and outputs an array of valid RSS feed items. (See [more about writing glob patterns](/en/guides/imports/#glob-patterns) for specifying which pages to include.)
```

> You can apply `pagesGlobToRssItems()` like so. This function assumes that you are globbing for items inside `src/pages/`, and all necessary feed properties are present in each document's frontmatter.

Reviewer: ```suggestion
This function assumes, but does not verify, that all necessary feed properties are present in each document's frontmatter. If you encounter errors, verify each page frontmatter manually.
```
Again, I think the examples are good enough to speak for themselves. So, we can use this as an opportunity to emphasize that this is DIY when it comes to making sure you have the required frontmatter properties. So something like this is maybe more helpful as the user-facing content?

```js title="src/pages/rss.xml.js" "pagesGlobToRssItems" "await pagesGlobToRssItems("
import rss, { pagesGlobToRssItems } from '@astrojs/rss';

export async function get(context) {
  return rss({
    title: 'Buzz’s Blog',
    description: 'A humble Astronaut’s guide to the stars',
    site: context.site,
    items: await pagesGlobToRssItems(
      import.meta.glob('./blog/*.{md,mdx}'),
    ),
  });
}
```

### Including full post content

<Since v="1.6.14" />

> Items also accept a `content` key containing the full content of the post as HTML.

Reviewer: ```suggestion
The `content` key contains the full content of the post as HTML. This allows you to make your entire post content available to RSS feed readers.
```
If we've mentioned the `content` property above, I think we can refer to it as a known/assumed thing in this intro.

:::tip
Whenever you're using HTML content in XML, suggest using a package like [`sanitize-html`](https://www.npmjs.com/package/sanitize-html) in order to make sure that your content is properly sanitized, escaped, and encoded.
:::

> When using content collections, you may render the post `body` using a standard Markdown parser like [`markdown-it`](https://github.com/markdown-it/markdown-it) and sanitizing the result:

Reviewer: ```suggestion
When using content collections, render the post `body` using a standard Markdown parser like [`markdown-it`](https://github.com/markdown-it/markdown-it) and sanitizing the result:
```

```js title="src/pages/rss.xml.js" ins={2, 3, 4, 14}
import rss from '@astrojs/rss';
import sanitizeHtml from 'sanitize-html';
import MarkdownIt from 'markdown-it';
const parser = new MarkdownIt();

export async function get(context) {
  const blog = await getCollection('blog');
  return rss({
    title: 'Buzz’s Blog',
    description: 'A humble Astronaut’s guide to the stars',
    site: context.site,
    items: blog.map((post) => ({
      link: `/blog/${post.slug}/`,
      content: sanitizeHtml(parser.render(post.body)),
      ...post.data,
    })),
  });
}
```

> When using glob imports with Markdown specifically, we suggest using the `compiledContent()` helper to retrieve the rendered HTML for sanitization:

Reviewer: ```suggestion
When using glob imports with Markdown, we suggest using the `compiledContent()` helper to retrieve the rendered HTML for sanitization:
```
The specifically just begs for more info. Is this Markdown files as opposed to MDX?  If so, I'd include a separate statement for what happens with MDX.

If it's as opposed to something else, what is that something else, and what do I need to do if I'm using that something else?

Author: It does indeed bork, but wanted to avoid saying it borks. I'll add a bork note 👍
